[driver] Adding Semtech SX1280/SX1281 driver

rasmuskleist opened the pull request:
Thanks @LasseJensen213 for contributing to this!

Before finishing up this driver I have a couple of points that I would like your opinion on:
1. Similar to [this](https://os.mbed.com/teams/Semtech/code/SX1280Lib/) implementation, I have added a hardware abstraction layer. This layer should allow using both spi and uart (yet to be implemented) similar to the transport layer in e.g. lis3. This layer also handles the differences between the command structure of the spi and uart interface, for example that the uart interface should also send the number of params. Furthermore, the register access and data buffer operations have different structures, so these functions have also been implmented in hal. Because this layer handles more than just mere transport I choose to call it hal. Currently I am not satisfied with the implmentation for the following reason:
    - I need to implement the uart hal, but in the current implemtation of modm the uart interface is somewhat different from both spi and i2c. In addition I have not been able to find a driver in modm, which uses uart. Is there any good reason why there is no resumable and device implementation of uart? If so how do you envision a driver using uart should be implemented?
    - I would like to have the sx128x struct members available in hal. However, if I simply let both Sx128x and Sx128xHalSpi inherit from sx128x I get the warning -Winaccessible-base and several compile errors e.g. 'modm::sx128x' is an ambiguous base. Meanwhile, if only Sx128xHalSpi inherit from sx128x then the members of sx128x are not available to Sx128x. Do you have any work around to this?
    - I am debating if I should simply have a transport layer which implements writeCommand and readCommand. These take a struct opcode and optionally offset or address. In addition it should take an array containing the parameters. These functions then figure out depending on the opcode if any of the optional parameters should be printed. What do you think about this?
   
1. Currently the definitions contains many duplicates in different name spaces. The idea behind this implementation is to make it very clear what definitions are available to each different packet engine. However, I think this implementation has two big problems. First of all it involves many nested namespaces so accessing members becomes very verbose. Second of all it makes it more difficult to maintain with code duplication. Two other options are:
     - The enum classes can be in the namespace sx128x and joined. With this implmenetation some options may not be applicable and a number may have different meanings. For example the bitrates are different between GFSK and FLRC and the value 0x45 means Br1000Bw1200 for GFSK and Br1300Bw1200 for FLRC. Hence only a subset of the resulting bitrate enum class would be applicable to GFSK and FLRC, respetively. I could specifiy which are applicable with comments.
     - The enum classes can be in the namespace sx128x and named according to which packet engine it applies to. This is implemented [here](https://os.mbed.com/teams/Semtech/code/SX1280Lib/). For example the bitrates are shared between GFSK and BLE, menawhile it is different for FLRC. The set of bitrates applicable to GFSK and BLE would then be called something like BitrateGfskBle and BitrateFlrc, respectively.

Thanks in advance for your comments, I really appreciate it! I should add that I have tested that I can send and recieve packets with LoRa.

salkinium commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,279 @@
+        Exti::connect<Dio2>(Exti::Trigger::RisingEdge, [](uint8_t) mutable { rx::dio2 = true; });
+
+        Dio3::setInput(modm::platform::Gpio::InputType::PullDown);
+        Exti::connect<Dio3>(Exti::Trigger::RisingEdge, [](uint8_t) mutable { rx::dio3 = true; });
```
```suggestion
        Exti::connect<Dio3>(Exti::Trigger::RisingEdge, [](uint8_t) { rx::dio3 = true; });
```
I think mutable only refers to objects inside the closure, but here you're modifying s]global objects, so it's not needed.

rasmuskleist replied:
I have removed the mutable keyword

salkinium commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,279 @@
+        Exti::connect<Dio2>(Exti::Trigger::RisingEdge, [](uint8_t) mutable { tx::dio2 = true; });
+
+        Dio3::setInput(modm::platform::Gpio::InputType::PullDown);
+        Exti::connect<Dio3>(Exti::Trigger::RisingEdge, [](uint8_t) mutable { tx::dio3 = true; });
```
```suggestion
        Exti::connect<Dio3>(Exti::Trigger::RisingEdge, [](uint8_t) { tx::dio3 = true; });
```
Same

salkinium commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,279 @@
+                    MODM_LOG_DEBUG << "Counter: " << counter << modm::endl;
+                    counter++;
+                }
+                tx::dio1 = false;
```
It's probably fine for the example, but this way you might miss an interrupt if it occurs before this line.
Making the `dio` into `std::atomic_bool` (no volatile) and then using `if (tx::dio.exchange(false)) {...}` will be interrupt safe.

rasmuskleist replied:
That is a very good point! This could lead to some interesting behaviour in my own use case, so thanks for pointing it out. I have changed it to use `std::atomic_bool`.

salkinium commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,219 @@
+{
+
+/**
+ * @tparam	Hal	    Hardware Abstraction Layer for SX128x
```
Technically a Transport or Protocol Layer ;-P
HAL typically refers to the abstraction of the microcontroller peripherals, like UART/SPI/I2C.

rasmuskleist replied:
Yeah I see your point! I took the naming from this [repo](https://os.mbed.com/teams/Semtech/code/SX1280Lib/). I have changed the name to transport.

salkinium commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,219 @@
+    modm::ResumableResult<bool>
+    getStatus(Status &status);
+
+    /// Writes a block of bytes in a data memory space starting at a specific address
```
does it?

rasmuskleist replied:
I have not fully implemented getStatus yet, because its command structure does not include a wait cycle as all others in SPI. I will return to this when finally having settled on how much responsibility the transport layer should have.

salkinium commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,219 @@
+    modm::ResumableResult<bool>
+    writeRegister(Register reg, uint8_t *data, size_t length);
+
+    /// Read a block of data starting at a given address
```
nope

rasmuskleist replied:
What?

salkinium replied:
A single byte of data. But its entirely copied from the function below, which is different

rasmuskleist replied:
Right, I have fixed this now.

salkinium commented on src/modm/driver/radio/sx128x_hal.hpp:
```diff
@@ -0,0 +1,92 @@
+{
+
+/**
+ * SX128X SPI Hardware Abstraction Layer.
```
```suggestion
 * SX128X SPI Transport Layer.
```
or Hardware Abstraction Abstraction Layer (HAAL)

salkinium commented on src/modm/driver/radio/sx128x_hal_impl.hpp:
```diff
@@ -0,0 +1,202 @@
+    Reset::setInput();
+    modm::delay_ms(20);
+
+    return true;
```
This is going to block the whole system for ~70ms, is that ok?

rasmuskleist replied:
I was debating if I should use a timeout instead. I believe resets would be rare (only upon initialization) so in that aspect it would be ok. However, I will change it to a timeout and use `RF_WAIT_UNTIL(timeout.isExpired())` instead. I have not implmeneted this yet because I will probably put this in Sx128x instead of the transport layer.

salkinium commented on src/modm/driver/radio/sx128x_impl.hpp:
```diff
@@ -0,0 +1,412 @@
+template < class Hal >
+Sx128x< Hal >::Sx128x()
+{
+}
```
Don't think this is necessary.

rasmuskleist replied:
I have changed it to use the default constructor for now

salkinium commented:
(same thing with the &arg-> *arg conversion here)

> Because this layer handles more than just mere transport I choose to call it hal

Please call it anything else but HAL, it has a predefined meaning in embedded software and it's not this. Maybe Protocol Layer or something.

> Is there any good reason why there is no resumable and device implementation of uart? If so how do you envision a driver using uart should be implemented?

UART has no state like SPI CS or I2C Transactions, it can only send packets. The protocol on top needs to be implemented by the caller. The async interface is basically `RF_WAIT_UNTIL(Uart::write(byte))`. Note that the buffers are off by default, you need to enabled them via an lbuild option. You can static assert their minimal size in your protocol driver though.

> I have not been able to find a driver in modm, which uses uart. 

That's because there are few drivers that use UART due to requiring a higher level framing, which is annoying to implement in an ASIC without CPU. I2C/SPI are much simpler in that respect.
You can, however, look at the very complicated [AMNB protocol driver](https://github.com/modm-io/modm/blob/develop/src/modm/communication/amnb/interface.hpp#L67-L97), which uses UART very extensively.

> if I simply let both Sx128x and Sx128xHalSpi inherit from sx128x I get the warning -Winaccessible-base and several compile errors e.g. 'modm::sx128x' is an ambiguous base.

The inheritance is only a convenience for the developer.
I would only inherit `modm::Sx128x` from `modm::sx128x` and use the full qualifiers or a typedef for Sx128xHalSpi.

> I should simply have a transport layer which implements writeCommand and readCommand.

Sure, whatever is the minimal set of function that need to be ported between SPI and UART. It's not an exposed API to the end user so it doesn't need to have all the syntax sugar of MODM_FLAGS for example.

> Currently the definitions contains many duplicates in different name spaces.

Yes it's annoying to write and maintain, but the other options are not necessarily easier to maintain either. I think its fine with the union as it is now.

rasmuskleist commented:
Thanks for the nice comments!

I have changed references to pointers. I will try to look into the AMNB protocol driver. 

Also I have been debating if I should make the transport layer much light so it is mere transport. Specifically, I was considering if I should declare a `Sx128xTransport` which defines the interface and a `Command` struct which contains the opcode, and vargs. The vargs would be empty for all opcodes expect register access and buffer operation commands. Thereby I can make the transport layer first write the contents of the `Command` struct followed by a waitCycle for the case of SPI read or the length for the case of UART read/write. I believe this will also make it possible to handle special cases such as getStatus, which is the only read operation that that does not have a waitCycle with SPI. There may be smarter ways to go about this, or what do you think?

chris-durand commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,215 @@
+
+    /// Writes a block of bytes in a data memory space starting at a specific address
+    modm::ResumableResult<bool>
+    writeRegister(Register reg, uint8_t *data, size_t length);
```
`data` should be `const` here and all the other `write*` functions. We could also consider using `std::span` instead of pointer + length arguments in new code.

rasmuskleist replied:
True I was considering if I should have done that but decided not to in respect of what is currently done in most other drivers. So for write this should be std::span<const uint8_t> and for read std::span<uint8_t>?

chris-durand replied:
`std::span<const uint8_t>` for `const` and `std::span<uint8_t>` for non-`const`.

rasmuskleist commented:
I should have addressed most issues by now I believe. What is now missing are the following

- [x] Implement (non-blocking) reset function
- [x] Fully implement UART transport layer
- [x] Implement and streamline logic in driver functions

With regards to the reset function I am debating if the driver should take a third template argument, the reset pin, or if the reset function should take the reset pin as a template argument? At the moment I cannot imagine calling the reset function anywhere inside the driver.

For the UART transport layer, I am considering if it makes sense to add UartDevice to architecture? The DeviceWrapper in AMNB seem to be very generic and something that can be reused for example in the sx128x driver. What are your thoughts on this?

Finally, I am debating how much logic should be implemented in the driver to ensure correct usage. After all it is a fairly complicated driver, so covering every possible inccorrect use case may be quite involved. At the moment I have kept the driver fairly light weight and have put the responsiblity of correct usage onto the user. In fact the only logic that I have implemented so far is to clear the interrupts before setting TX and RX. However, I think it may be better to add a warning that the interrupts should be cleared before setting TX or RX and leave it up to the user. This is because I imagine every use case to manually clear the interrupt in the protothread as in the example. What is your viewpoint on this?

salkinium commented:
> if the driver should take a third template argument

Probably yes, cos that's how all other drivers do it.

> I am considering if it makes sense to add UartDevice to architecture? 

You mean a wrapper that provides generic transaction-like semantics?

1. write
2. wait for write finish
3. read
4. wait for read finish

Yeah, I can see that being a useful pattern for a bunch of UART-based drivers.

>  I am debating how much logic should be implemented in the driver to ensure correct usage. After all it is a fairly complicated driver, so covering every possible inccorrect use case may be quite involved

It's probably best to first understand the typical usage and then apply the error detection/recovery for those paths first.
Otherwise you're perhaps implementing something that isn't even how people (ab)use it in the real world.
The driver doesn't need to be perfect, you can always come back and fix it later if you find that some things annoy you.

rasmuskleist commented:
I should have now fully implemented the UART transport layer. What is now missing is to handle some special cases where the read value from the radio should be manipulated e.g. the rssi. Typically this is done in for example a Data class in modm, which is nice because it keeps the driver lean. However, for the PacketStatus struct this would mean having to implement duplicate getter method for each PacketType, a lot of which are duplicates. Therefore it may be advantageous to implement this conversion inside the driver instead. Especially to keep things consistent because the getRssi function in the driver should do the conversion as well. What are your thoughts on this?

salkinium commented on examples/nucleo_h723zg/adc_injected_conversion/main.cpp:
Don't think these files belong in here

salkinium commented on src/modm/architecture/interface/uart_device.hpp:
```diff
@@ -0,0 +1,137 @@
+ * @author	Rasmus Kleist Hørlyck Sørensen
+ * @ingroup modm_architecture_uart_device
+ */
+template < class Uart, uint8_t NestingLevels = 10, uint16_t TxUsTimeout = 1000, uint16_t RxUsTimeout = 10000 >
```
I would perhaps make the timeouts into member variables. It's a bit unclean to pass them as template params and maybe you need to adapt them at runtime. (The AMNB code needs to be refactored).

rasmuskleist replied:
I took the inspiration from AMNB, but yeah it makes sense to have them as member variables.

salkinium commented on src/modm/driver/radio/sx128x_transport.hpp:
```diff
@@ -0,0 +1,149 @@
+    writeCommand(Command command, std::span<const uint8_t> data) = 0;
+
+    virtual modm::ResumableResult<bool>
+    readCommand(Command command, std::span<uint8_t> data) = 0;
```
Don't think these virtual methods are necessary, since you're passing the specialized class as a template parameter anyways, so the type is fully known. You should be able to remove these definitions and it should still work.

rasmuskleist replied:
Thats right! It was only to layout the transport interface, but I have removed these.

salkinium commented on src/modm/platform/adc/stm32f0/adc_impl.hpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/adc/stm32f3/adc.hpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/adc/stm32f3/adc_impl.hpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/adc/stm32f3/module.lb:
This file shouldn't be in here.

salkinium commented on src/modm/platform/can/stm32-fdcan/can.cpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/can/stm32-fdcan/can.hpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/can/stm32-fdcan/can_shared_irqs.cpp.in:
This file shouldn't be in here.

salkinium commented on src/modm/platform/can/stm32-fdcan/module.lb:
This file shouldn't be in here.

salkinium reviewed (commented):
Very thorough driver, almost done, right?

> What are your thoughts on this?

You can remove the data class completely if you want. It's a bad convention anyways. Then you don't need to duplicate configuration data everywhere.

(There appears to have been some mishap with the rebase, there are commits in here that are already in develop.)

rasmuskleist commented:
Yes I believe its done, at least for now. As you write it is very difficult to predict how people (including myself) will (ab)use the driver at this point. Therefore I would rather do some changes at later to make it easier to use. I presume that I can do some breaking changes at later with this driver if necessary? I also decided to dumb the driver down, so the driver only facilitates a nice interface for issuing the commands. This is for simplicity and consistency.

I will confirm that the changes made to the driver physically works tonight and convert to PR. Also I should add that I discovered the EXTI bug to be resolved with #1052, when testing this driver. It caused a lot of confusion so I am happy to see that there is already a fix for it!

I don't know how I messed that rebase up, but it should be fixed now.

rasmuskleist committed:
[architecture] Adding UartDevice for resumable r/w

rasmuskleist committed:
[driver] Adding Semtech SX1280/SX1281 driver

rasmuskleist committed:
[example] Adding SX128x LoRa example

chris-durand commented:
If you need the EXTI fixes from #1052, you can cherry-pick the commits and add them to this PR. I probably won't find any time at work this week to finish my PR.

I'll have a look at this driver tomorrow.

chris-durand commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,277 @@
+static constexpr modm::sx128x::LoRa::ModulationParams modulationParams = {
+    spreadingFactor : modm::sx128x::LoRa::ModulationParams::SpreadingFactor::Sf9,
+    bandwidth : modm::sx128x::LoRa::ModulationParams::Bandwidth::Bw400,
+    codingRate : modm::sx128x::LoRa::ModulationParams::CodingRate::Cr_Li_4_7
```
I have never seen this syntax with `:` for designated initializers :exploding_head:. According to the [GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html) this syntax is a compiler extension for pre-standard C considered obsolete since GCC 2.5. Could you use the C++ standard syntax (`.x = `) instead?

rasmuskleist replied:
Thanks for pointing that out! I believe I have the syntax `:` lurking multiple places in application code...

chris-durand commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,277 @@
+    headerType : modm::sx128x::LoRa::PacketParams::HeaderType::Explicit,
+    payloadLength : 4,
+    crc : modm::sx128x::LoRa::PacketParams::Crc::Enable,
+    invertIq : modm::sx128x::LoRa::PacketParams::InvertIq::Standard
```
same as above

chris-durand commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,277 @@
+                    /// TODO: read frequency error indicator (FEI) from register 0x0954 (MSB) 0x0955, 0x0956 (LSB).
+
+                    MODM_LOG_DEBUG << "Received Message" << modm::endl;
+                    MODM_LOG_DEBUG << "Counter: " << reinterpret_cast<uint32_t*>(buffer)[0] << modm::endl;
```
Casting the `uint8_t` pointer to `uint32_t*` and dereferencing it is undefined behaviour, see https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing. The opposite cast would work, though.

rasmuskleist replied:
Right, I have changed it so I cast a `uint32_t` to `*uint8_t` and use `std::memcpy`. I believe that should do?

chris-durand commented on examples/nucleo_g474re/sx128x_lora/main.cpp:
```diff
@@ -0,0 +1,277 @@
+
+                    PT_CALL(radio.getRxBufferStatus(&rxBufferStatus));
+                    PT_CALL(radio.getPacketStatus(&packetStatus));
+                    PT_CALL(radio.readBuffer(rxBufferStatus.rxStartBufferPointer, std::span{buffer, rxBufferStatus.rxPayloadLength}));
```
Is it guaranteed `rxBufferStatus.rxPayloadLength` is never larger than `bufferSize`?

rasmuskleist replied:
I checked that the transceiver is equipped with a 256 byte RAM data buffer, so I will change the buffer size in the thread to be the same. Then there should be no problem.

chris-durand commented on src/modm/architecture/interface/uart_device.hpp:
```diff
@@ -0,0 +1,156 @@
+		timeout.restart(rxTimeout);
+		while (readIndex < length)
+		{
+			if (size_t writeSize = Uart::read(&buffer[readIndex], length - readIndex))
+			{
+				writeIndex += writeSize;
```
Shouldn't this be `readSize` and `readIndex`? Looks like a copy-paste error from `write()`. The loop will never terminate if `length != 0` because `readIndex` is never incremented.

rasmuskleist replied:
Yes copy-pasting gone wrong... I should add that I have not tested the UartDevice interface and hence why I have not caught this mistake, that would prevail while reading.

chris-durand commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,247 @@
+    Sx128x() = default;
+
+    /// Determine if radio is busy
+    inline bool
```
```suggestion
    bool
```
`inline` is not needed here because the class is a template.

chris-durand commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,247 @@
+    reset();
+
+public:
+    /// Get the transciever status directly.
```
```suggestion
    /// Get the transceiver status directly.
```
spelling mistake

chris-durand commented on src/modm/driver/radio/sx128x.hpp:
```diff
@@ -0,0 +1,247 @@
+    modm::ResumableResult<bool>
+    writeBuffer(uint8_t offset, std::span<const uint8_t> data);
+
+    /// This function allows reading (n-3) bytes of payload received starting at offset
```
What is `n` in the comment?

rasmuskleist replied:
Right that is a bit confusing. I believe I have written this because the full message of size `n` has three command bytes (opcode, offset, nop) and thus reads `n-3`. I have changed this comment since the user should not care about the framing of the message.

chris-durand reviewed (commented):
Thank you!

rasmuskleist commented:
Thank you for your comments! I should have addressed your comments now. With regards to your fix I can easily wait until your PR is ready. At the moment I will just use the device as if there was a single interrupt pin by issuing getIrq.

salkinium reviewed (approved):
Thanks for another great driver!

salkinium commented:
(It would be good if you could resolve the conversations yourself, when you feel like you addressed them. Then we don't have to check again and it's clear to us what the state is. We can always unresolve them if something comes up.)

rasmuskleist commented:
Thanks! Yes I will do that going on.
